apps/rewards/src/invite/inviteReward.service.ts

```diff
@@ -0,0 +1,171 @@
+import { KomenciLoggerService } from '@app/komenci-logger'
+import { networkConfig, NetworkConfig } from '@app/utils/config/network.config'
+import { isAccountConsideredVerified } from '@celo/base/lib'
+import { CeloContract, ContractKit } from '@celo/contractkit'
+import { Inject, Injectable } from '@nestjs/common'
+import { Cron, CronExpression } from '@nestjs/schedule'
+import { Raw } from 'typeorm'
+import { v4 as uuidv4 } from 'uuid'
+import { NotifiedBlock } from '../blocks/notifiedBlock.entity'
+import { NotifiedBlockRepository } from '../blocks/notifiedBlock.repository'
+import { InviteReward, RewardStatus } from './inviteReward.entity'
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEEKLY_INVITE_LIMIT = 20
+
+@Injectable()
+export class InviteRewardService {
+  isRunning = false
+  komenciAddresses = []
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly notifiedBlockRepository: NotifiedBlockRepository,
+    private readonly contractKit: ContractKit,
+    @Inject(networkConfig.KEY)
+    private readonly networkCfg: NetworkConfig,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.komenciAddresses = this.networkCfg.relayers.map(
+      relayer => relayer.externalAccount
+    )
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async sendInviteRewards() {
+    if (this.isRunning) {
+      this.logger.log('Skipping because previous run is still ongoing')
+      return
+    }
+    this.isRunning = true
+
+    const cUsdTokenAddress = (
+      await this.contractKit.registry.addressFor(CeloContract.StableToken)
+    ).toLowerCase()
+
+    await this.usingLastNotifiedBlock(async fromBlock => {
+      this.logger.log(
+        `Starting to fetch invite rewards from block ${fromBlock}`
+      )
+      const escrow = await this.contractKit.contracts.getEscrow()
+      const withdrawalEvents = await escrow.getPastEvents('Withdrawal', {
+        fromBlock
+      })
+      this.logger.log(`Withdrawal events received: ${withdrawalEvents.length}`)
+      let maxBlock = fromBlock
+      for (const withdrawal of withdrawalEvents) {
+        const {
+          transactionHash,
+          blockNumber,
+          returnValues: { identifier, token, to }
+        } = withdrawal
+        maxBlock = Math.max(maxBlock, blockNumber)
+
+        const inviter = to.toLowerCase()
+        if (cUsdTokenAddress !== token.toLowerCase()) {
+          continue
+        }
+        const tx = await this.contractKit.web3.eth.getTransaction(
+          transactionHash
+        )
+        if (!this.isKomenciSender(tx.from)) {
+          continue
+        }
+        const invitee = tx.to?.toLowerCase() ?? ''
```

Comment:
Will `tx.to` return the recipient of the ERC20 Transfer? It seems that this `tx.to` would be the address of the Escrow contract (and not the msg.sender)

Reply:
We might be able to get this by looking at the events emitted from the transaction:
https://ethereum.stackexchange.com/questions/66454/how-to-get-events-emitted-by-a-transaction-with-web3-js/66459

Keep in mind that we don't want to pay the withdrawer (which will be the users' Metatransaction Wallet). After the funds are withdrawn to the MTW, they're transferred to the Externally Owned Account. See here:
https://github.com/celo-org/wallet/blob/7a5d896055f519e695f9b51a0d1aae98dcb3da79/packages/mobile/src/escrow/saga.ts#L370

Reply:
I feel like this two comments are referring to different things?

For the first one:
This `tx.to` that we're using for invitees surprised me a bit too but it belongs to MTW contract addresses. Here are some examples on Alfajores:
```
0x3862bc669c12ed779b2258c6fef4435c1ce860fc
0x42a31ae4b415296661b9294514ca130bedeff908
0x458455c4a79f383bc5561544fceaaf09606d6cf5
0x5e16845cba46570883d03f63153fe1ed0dd6a270
0xb11e9aac1d529d7c2e5d9d97c13c1f42923fac8d
```

Those are the ones being verified so that's the one we're using for the `addressIsConsideredVerified` function which looks to be working as well.

For the `inviter` (the one we're going to be paying) we're using the `to` field of the `Withdrawal` event, which is fetched from the `EscrowedPayment` that's created in the `transfer` function:
https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/identity/Escrow.sol#L130 which looks like they are EOA addresses. Some examples from Alfajores:
```
0xdda9eea8354451e645ea91fca32b6bdf387a8fc0
0xcab02027849e5fad820cfd3368cb0d86fb25face
0x0f3c917a23539dd072dd0c36e75926f05567e4b8
0x65ebd8f87f1534003943934f9aba5104f9685421
```

I might be missing something here, so please let me know if I misunderstood your comment :)

Reply:
> This tx.to that we're using for invitees surprised me a bit too but it belongs to MTW contract addresses.

Ah I see, I was thinking that we'd be paying this address and wanted to make sure we payed the EOA instead. Makes sense that you're getting this to verify the MTW is considered verified and that we're actually paying the inviter.

Reply:
In the case that `tx.to` isn't defined, we should log error and `continue` right?

Reply:
Yes. I figured that it would not pass the verified account check, but it's better to avoid the extra unnecessary queries.